Source/WTF/wtf/text/StringConcatenate.h

```diff
@@ -250,6 +280,44 @@ template<> class StringTypeAdapter<AtomStringImpl&, void> : public StringTypeAda
     }
 };
 
+struct FromUTF8 {
+    const char* characters;
+    unsigned lengthUTF8 { 0 };
+    unsigned lengthUTF16 { 0 };
+    bool is8Bit { true };
+    bool conversionFailed { true };
+
+    FromUTF8(const char* characters, size_t length)
+        : characters { characters }
+    {
+        auto result = Unicode::computeUTFLengths(characters, characters + length);
+        if (result.result == Unicode::ConversionResult::SourceIllegal)
+            return;
+        if (result.lengthUTF16 > String::MaxLength)
+            return;
+        lengthUTF8 = result.lengthUTF8;
+        lengthUTF16 = result.lengthUTF16;
+        is8Bit = result.isAllASCII;
+        conversionFailed = false;
+    }
+};
+
+template<> class StringTypeAdapter<FromUTF8, void> {
+public:
+    StringTypeAdapter(FromUTF8 characters)
```

Comment:
Since the object has all scalars in it, I think it’s fine to pass it by value. It would be OK to write out the more verbose `const&` form, but I think there is no benefit.